src/WebJobs.Extensions.SendGrid/Config/SendGridConfiguration.cs

```diff
@@ -0,0 +1,126 @@
+        /// </remarks>
+        public EmailAddress ToAddress { get; set; }
+
+        /// <summary>
+        /// Gets or sets the default "from" address that will be used for messages.
+        /// This value can be overridden by job functions.
+        /// </summary>
+        public EmailAddress FromAddress { get; set; }
+
+        internal ISendGridClientFactory ClientFactory { get; set; }
+
+        /// <inheritdoc />
+        public void Initialize(ExtensionConfigContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            var metadata = new ConfigMetadata();
+            context.ApplyConfig(metadata, "sendgrid");
+            this.ToAddress = SendGridHelpers.Apply(this.ToAddress, metadata.To);
+            this.FromAddress = SendGridHelpers.Apply(this.FromAddress, metadata.From);
+
+            if (string.IsNullOrEmpty(this.ApiKey))
+            {
+                INameResolver nameResolver = context.Config.NameResolver;
+                this.ApiKey = nameResolver.Resolve(AzureWebJobsSendGridApiKeyName);
+            }
+
+            context                
+                .AddConverter<string, SendGridMessage>(SendGridHelpers.CreateMessage)
+                .AddConverter<JObject, SendGridMessage>(SendGridHelpers.CreateMessage)
+                .AddBindingRule<SendGridAttribute>()
+                .AddValidator(ValidateBinding)
+                    .BindToCollector<SendGridMessage>(CreateCollector);
+        }
+
+        private IAsyncCollector<SendGridMessage> CreateCollector(SendGridAttribute attr)
+        {
+            string apiKey = FirstOrDefault(attr.ApiKey, ApiKey);
+            ISendGridClient sendGrid = _sendGridClientCache.GetOrAdd(apiKey, a => ClientFactory.Create(a));
+            return new SendGridMessageAsyncCollector(this, attr, sendGrid);
+        }
+
+        private void ValidateBinding(SendGridAttribute attribute, Type type)
+        {
+            ValidateBinding(attribute);
+        }
+
+        private void ValidateBinding(SendGridAttribute attribute)
+        {
+            string apiKey = FirstOrDefault(attribute.ApiKey, ApiKey);
+
+            if (string.IsNullOrEmpty(apiKey))
+            {
+                throw new InvalidOperationException(
+                    $"The SendGrid ApiKey must be set either via an '{AzureWebJobsSendGridApiKeyName}' app setting, via an '{AzureWebJobsSendGridApiKeyName}' environment variable, or directly in code via {nameof(SendGridConfiguration)}.{nameof(SendGridConfiguration.ApiKey)} or {nameof(SendGridAttribute)}.{nameof(SendGridAttribute.ApiKey)}.");
+            }
+        }
+
+        private static string FirstOrDefault(params string[] values)
+        {
+            return values.FirstOrDefault(v => !string.IsNullOrEmpty(v));
+        }
+
+        // Schema for host.json 
+        private class ConfigMetadata
```

Comment:
think we should also support fromAddress and toAddress in json

Reply:
You mean just add additional JsonProperty annotations for those alternate names?

Reply:
Yeah, I think that might do the trick